lib/SILOptimizer/Mandatory/Differentiation/DerivativeLookup.cpp

```diff
@@ -0,0 +1,122 @@
+//===--- DerivativeLookup.cpp ---------------------------------------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2019 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+//
+// SWIFT_ENABLE_TENSORFLOW
+//
+// Utilities for looking up derivatives of functions.
+//
+//===----------------------------------------------------------------------===//
+
+#include "DerivativeLookup.h"
+
+namespace swift {
+
+/// Returns the AbstractFunctionDecl corresponding to `F`. If there isn't one,
+/// returns `nullptr`.
+static AbstractFunctionDecl *getAFDOrNull(SILFunction *F) {
```

Comment:
Very minor:
```suggestion
static AbstractFunctionDecl *getAbstractFunctionDeclOrNull(SILFunction *F) {
```
I'd prefer no empty lines for small helper functions, if you don't mind.

Reply:
To declare a helper that finds an AST node given a SIL construct, I usually use `find*` (e.g. `findDifferentialOperator` in Differentiation.cpp). In general, in Differentiation.cpp my preference is to use `find*` if the result is nullable, and `get*` if the result is non-null.

Reply:
sounds good

Reply:
Done in https://github.com/apple/swift/pull/28451/commits/bb932588856e23b74417368148bc84b0891b3252.